### DISCUSSION TITLE
Function to await keyboard shortcuts / chords

### DISCUSSION THREAD

**Discussion by ggoodman** [johnlindquist/kit#133] [Ideas]
## Description

Scriptkit already exposes the `arg` function as a mechanism to await user-provided input and may benefit from a similar paradigm for accepting user-provided keyboard shortcuts.

## Use-case

I would like to create a script that:

1. Enters a 'layout mode'. When the script is in layout mode, it is awaiting user inputs that may affect the positioning of the current window.
2. While in 'layout mode', blocks (waits on) one of a set of keyboard shortcuts or chords. Key presses and chords that are not specified are ignored.
3. Do some pattern matching based on the actual hotkey / chord received, or obtain a logical identifier for such. This pattern matching will make it easy to:
  - Break out of the 'layout mode' (ie: `esc`).
  - Determine which layout logic to apply to the active window.

## Sample code

```js

const keybindings = {
  left: ['Cmd-Left'],
  right: ['Cmd-Right'],
  exit: ['Escape'],
};

let positioning = true;

while (positioning) {
  const pressed = await hotkey(keybindings);

  switch (pressed.name) {
    case 'left': positionWindowOnLeft(); break;
    case 'right': positionWindowOnRight(); break;
    case 'exit': positioning = false; break;
  }
}
```

**Comment by johnlindquist** [comment]
OOOooohh, I _love_ this idea and your implementation looks spot-on. 💪

> Fyi, semi-tangentially-related there is an `exit()` fn to end a script early

This concept will probably even help with the case of assigning keyboard shortcuts to scripts through the UI.

I'll give it a go this week. Got a window resizing task I need to focus on today/tomorrow.

Thanks so much for the idea/concept!

**Comment by johnlindquist** [comment]
Here's the progress I made with this today. Going to try to push out a build late tonight

https://user-images.githubusercontent.com/36073/114948758-e6912480-9e0c-11eb-9f3c-6ec870d3a4ea.mp4
